### DISCUSSION TITLE
How to develop on (slow) remote machines?

### DISCUSSION THREAD

**Discussion by rodja** [zauberzeug/nicegui#84] [Q&A]
One major driver behind the creation of NiceGUI was the need for a nice UI framework to control and configure robots and IoT devices. These are notoriously slow compared to development workstations and laptops. NiceGUI runs quite fine on such constrained hardware. But If you use [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) or similar to do actual development on these systems, everything feels like jelly. Especially if you run powerful extensions like [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance).

That's why we at [Zauberzeug](https://zauberzeug.com) created a small tool called [LiveSync](
https://github.com/zauberzeug/livesync). It combines a local filesystem watcher with rsync to copy changes to a (slow) remote target whenever your local code changes. This approach has multiple advantages:

- own choosing of IDE and tooling
- locally able to run tests (or the production code)
- simultaneously continuous deployment of the development code to the target environment (where auto-reload ensures live preview)
- almost no overhead on the (slow) target

So everyone: What do you think of this approach? Or how do you solve this kind of problems?
@me21 You mentioned developing for Beaglebone. I would love to read your opinion.

**Comment by me21** [comment]
I'm sorry, I didn't try your tool. I use Jetbrains PyCharm IDE for development, which provides a similar solution. On the bright side, it suggests you're on the right track and your approach is viable.
